### PR TITLE
Chart test zoom

### PR DESCRIPTION
**Pull request by cepprice**

### CODE REVIEW

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		}
+		double targetZoom = Math.min(tb.getZoom() + tb.getZoomFloatPart() + zoomDelta, settings.AUTO_ZOOM_MAP_SCALE.get().maxZoom);
+		targetZoom = Math.round(targetZoom * 3) / 3f;
+		int newIntegerZoom = (int) Math.round(targetZoom);
+		float zPart = (float) (targetZoom - newIntegerZoom);
+		return newIntegerZoom > 0 ? new ComplexZoom(newIntegerZoom, zPart) : null;
+	}
+
+	private static float defineZoomFromSpeed(@NonNull RotatedTileBox tb, float speed, float zoomProximityCoeff) {
+		if (speed < 7f / 3.6) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java:59]: this should be part of method for animated zoom

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+	public static ComplexZoom calculateAutoZoomBySpeed(@NonNull OsmandApplication app, @NonNull RotatedTileBox tb, float speed) {
+		OsmandSettings settings = app.getSettings();
+
+		float zoomProximityCoeff = settings.AUTO_ZOOM_MAP_SCALE.get().coefficient;
+		float zoomDelta = defineZoomFromSpeed(tb, speed, zoomProximityCoeff);
+		if (Math.abs(zoomDelta) < 0.5) {
+			return null;
+		}
+
+		if (zoomDelta >= 2) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java:46]: are zoom delta correct evaluations? what it means

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+import androidx.annotation.Nullable;
+
+public class AutoZoomBySpeedUtils {
+
+	public static final float ZOOM_PER_SECOND = 0.1f;
+	public static final int MIN_ZOOM_DURATION_MILLIS = 1500;
+	public static final long FIXED_ZOOM_DURATION_MILLIS = 1000;
+	private static final int SHOW_DRIVING_SECONDS = 60;
+
+	@Nullable
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java:36]: Too many static methods, class should be part of RoutingHelper or Location..Service or something similar and used by instance

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java (hunk 2 above)
Comment (cepprice) [reply OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedUtils.java:46]: It's moved code for V1

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+		float expectedSurfaceZoom = mapRenderer.getSurfaceZoomAfterPinch(
+				myLocation31, myLocationHeight, myLocationPixel,
+				anotherLocation31, anotherLocationHeight, anotherPixel);
+		if (expectedSurfaceZoom == -1.0f) {
+			return null;
+		}
+
+		int minZoom = mapView.getMinZoom();
+		int maxZoom = Math.min(mapView.getMaxZoom(), autoZoomScale.maxZoom);
+		Zoom boundedZoom = Zoom.checkZoomBounds(expectedSurfaceZoom, minZoom, maxZoom);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java:135]: this should be deleted

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+		float expectedSurfaceZoom = mapRenderer.getSurfaceZoomAfterPinch(
+				myLocation31, myLocationHeight, myLocationPixel,
+				anotherLocation31, anotherLocationHeight, anotherPixel);
+		if (expectedSurfaceZoom == -1.0f) {
+			return null;
+		}
+
+		int minZoom = mapView.getMinZoom();
+		int maxZoom = Math.min(mapView.getMaxZoom(), autoZoomScale.maxZoom);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java:134]: ```autoZoomScale.maxZoom``` should be deleted

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+		float showDistanceToDrive = filteredSpeed * SHOW_DRIVING_SECONDS_V2 / autoZoomScale.coefficient;
+		float rotation = rotationToAnimate != null ? rotationToAnimate : mapView.getRotate();
+		LatLon anotherLatLon = MapUtils.rhumbDestinationPoint(myLocationLatLon, showDistanceToDrive, rotation);
+
+		PointI anotherLocation31 = NativeUtilities.getPoint31FromLatLon(anotherLatLon);
+		float anotherLocationHeight = NativeUtilities.getLocationHeightOrZero(mapRenderer, anotherLocation31);
+		PointI windowSize = mapRenderer.getState().getWindowSize();
+		PointI anotherPixel = new PointI(windowSize.getX() / FOCUS_DIV_X, windowSize.getY() / FOCUS_DIV_Y);
+
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java:125]: Here you need to ```Math.max(measuredDist(	myLocation31, anotherLocation31),  autoZoomScale.minDistance)
```

File: OsmAnd/src/net/osmand/plus/views/Zoom.java [hunk 7]
```diff
@@ -100,9 +100,85 @@ private void checkZoomBounds() {
...
+		float zoomFactorFromStartIntZoom = (1 << (int) Math.abs(zoomDeltaFromStartZoomBase))
+				* (Math.abs(zoomDeltaFromStartZoomBase - (int) zoomDeltaFromStartZoomBase) + 1.0f);
+		if (zoomDeltaFromStartZoomBase < 0.0f) {
+			zoomFactorFromStartIntZoom = 1.0f / zoomFactorFromStartIntZoom;
+		}
+
+		return distanceNoStartZoomFloatPart / zoomFactorFromStartIntZoom;
+	}
+
+	public static float fromDistanceRatio(double startDistance, double endDistance, float startZoom) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/Zoom.java:130]: Good naming and example / explanation is needed

File: OsmAnd/src/net/osmand/plus/views/Zoom.java [hunk 8]
```diff
@@ -100,9 +100,85 @@ private void checkZoomBounds() {
...
+	}
+
+	@NonNull
+	public static Zoom checkZoomBounds(int baseZoom, float zoomFloatPart, int minZoom, int maxZoom) {
+		Zoom zoom = new Zoom(baseZoom, zoomFloatPart, minZoom, maxZoom);
 		zoom.checkZoomBounds();
 		return zoom;
 	}
+
+	public static float getZoomedDistance(float distance, float startZoom, float endZoom) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/Zoom.java:114]: Good naming and example / explanation is needed

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+			return null;
+		}
+
+		int minZoom = mapView.getMinZoom();
+		int maxZoom = Math.min(mapView.getMaxZoom(), autoZoomScale.maxZoom);
+		Zoom boundedZoom = Zoom.checkZoomBounds(expectedSurfaceZoom, minZoom, maxZoom);
+		return new ComplexZoom(boundedZoom.getBaseZoom(), boundedZoom.getZoomFloatPart());
+	}
+	@Nullable
+	public Pair<ComplexZoom, Long> getAutoZoomParams(float currentZoom, @NonNull ComplexZoom autoZoom, boolean fixedDuration) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java:194]: Here pass ```long fixedDuration```
```
if (fixedDuration > 0) {
			return new Pair<>(autoZoom, fixedDuration);
		}
```

File: OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+
+		OsmandMapTileView mapView = app.getOsmandMap().getMapView();
+		AutoZoomMap autoZoomScale = settings.AUTO_ZOOM_MAP_SCALE.get();
+
+		LatLon myLocationLatLon = new LatLon(myLocation.getLatitude(), myLocation.getLongitude());
+		PointI myLocation31 = NativeUtilities.getPoint31FromLatLon(myLocationLatLon);
+		float myLocationHeight = NativeUtilities.getLocationHeightOrZero(mapRenderer, myLocation31);
+		PointI myLocationPixel = mapRenderer.getState().getFixedPixel();
+
+		float showDistanceToDrive = filteredSpeed * SHOW_DRIVING_SECONDS_V2 / autoZoomScale.coefficient;
```
[the remark this answers is not in the document]
Comment (vshcherb) [reply OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java:117]: showDistanceToDrive = Math.max(Math.min(distanceToNextTurn, showDistanceToDrive),  autoZoomScale.minDistance)

[the remark this answers is not in the document]
Comment (vshcherb) [reply OsmAnd/src/net/osmand/plus/views/AutoZoomBySpeedHelper.java:117]:
```
distanceToNextTurn = DistanceToNextVisibleTurn();
```

**Pull request merged**